lib/utils/batcher/batcher.go

```diff
@@ -0,0 +1,172 @@
+func Run[T any](ctx context.Context, events <-chan T, fn func(batch []T) error, opts ...Option) error {
+	collector := New[T](opts...)
+	return trace.Wrap(collector.Run(ctx, events, fn))
+}
+
+// Option configures a Collector using the functional options pattern.
+type Option func(*Collector[any])
+
+// WithWindow sets the collection window duration.
+func WithWindow(d time.Duration) Option {
+	return func(c *Collector[any]) {
+		if d > 0 {
+			c.window = d
+		}
+	}
+}
+
+// WithThreshold sets the maximum number of events per batch.
+// Default is unlimited (bounded only by window duration).
+func WithThreshold(n int) Option {
+	return func(c *Collector[any]) {
+		if n > 0 {
+			c.threshold = n
+		}
+	}
+}
+
+// WithClock sets a custom clock for testing.
+// Default uses the system clock.
+func WithClock(clock clockwork.Clock) Option {
+	return func(c *Collector[any]) {
+		if clock != nil {
+			c.clock = clock
+		}
+	}
+}
+
+// New creates a new Collector with the given options.
+func New[T any](opts ...Option) *Collector[T] {
+	c := &Collector[T]{
+		window:    10 * time.Second,
+		threshold: 50,
+		clock:     clockwork.NewRealClock(),
+	}
+	// Apply options (type conversion needed for generics)
+	for _, opt := range opts {
+		opt((*Collector[any])(c))
+	}
+	return c
+}
+
+// Collector batches events into time windows.
+type Collector[T any] struct {
+	window    time.Duration
+	threshold int
+	clock     clockwork.Clock
+}
+
+// Run starts collecting events and calls fn for each batch.
+// It blocks until ctx is canceled or an error occurs.
+func (c *Collector[T]) Run(ctx context.Context, events <-chan T, fn func(batch []T) error) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		batch, err := c.CollectBatch(ctx, events)
+		if err != nil {
+			return err
+		}
+
+		if len(batch) > 0 {
+			if err := fn(batch); err != nil {
+				return trace.Wrap(err)
+			}
+		}
+	}
+}
+
+// CollectBatch collects events into a batch until the time window elapses
+// or the threshold is reached, whichever comes first.
+// It blocks until at least one event is collected or ctx is canceled.
+func (c *Collector[T]) CollectBatch(ctx context.Context, events <-chan T) ([]T, error) {
+	// lazy initialization of timer channel.
+	// We don't want to return if there is no events
+	// and block at least until the first event arrives.
+	// Where <-timerC blocks the timerC is overwritten after
+	// detecting the first event.
+	var timerC <-chan time.Time
+	var batch []T
+
+	for {
+		select {
+		case <-ctx.Done():
+			return batch, ctx.Err()
+
+		case <-timerC:
+			return batch, nil
+
+		case event, ok := <-events:
+			if !ok {
+				// Channel closed, return what we have
+				return batch, nil
+			}
+			batch = append(batch, event)
+			if len(batch) == 1 {
+				// First event in the batch, start the timer
+				// To bock on no events arriving and at least return one event
+				timerC = c.clock.After(c.window)
```

Comment:
Do we know if the fake clock from clockwork actually follows the 1.23 semantics for channel and timer leaking?

Reply:
Yea, Probably it can be buggy and the safer way is to use timer reset approach. 

Thanks, aligned in c4feccbc22826db584b4802fed1b7ac8cc622616